backends/guest/common/read.c

```diff
@@ -121,59 +122,46 @@ int print_cert_info(crypto_x509_t *x509)
 int print_variables(const uint8_t *buffer, size_t buffer_size, const char *var_name)
 {
 	int rc;
-	ssize_t esl_data_size = buffer_size, cert_size;
-	size_t count = 0, offset = 0;
-	uint8_t *cert = NULL, *esl_data = (uint8_t *)buffer;
-	enum signature_type sig_type;
-	sv_esl_t *sig_list;
-	sv_esd_t *esd = NULL;
+	size_t esl_data_size = 0;
+	size_t count = 0;
 	crypto_x509_t *x509 = NULL;
 
-	while (esl_data_size > 0) {
-		// TODO: consider breaking this down into functions, to avoid these scope reductions
-		size_t esl_size, next_esl_size, sig_offset;
-		size_t signature_size;
-
-		if (esl_data_size < sizeof(sv_esl_t)) {
-			prlog(PR_ERR,
-			      "ERROR: ESL has %zd bytes and is smaller than an ESL (%zu bytes),"
-			      " remaining data not parsed\n",
-			      esl_data_size, sizeof(sv_esl_t));
-			break;
-		}
+	union {
+		const uint8_t *raw;
+		sv_esl_t *esl;
+	} curr_esl;
 
-		/* Get sig list */
-		sig_list = extract_esl_signature_list(buffer + offset, esl_data_size);
-		esl_size = sig_list->signature_list_size;
-		signature_size = cpu_to_le32(sig_list->signature_size);
-		sig_type = get_signature_type(sig_list->signature_type);
+	rc = next_esl_from_buffer(buffer, buffer_size, &curr_esl.raw, &esl_data_size);
+	if (rc) {
+		prlog(PR_ERR, "Error reading from esl buffer: %d\n", rc);
+		return rc;
+	}
 
-		if (esl_size < sizeof(sv_esl_t) || esl_size > esl_data_size) {
-			prlog(PR_ERR, "ERROR: invalid ESL size (%zu)\n", esl_size);
-			break;
+	while (curr_esl.esl != NULL) {
+		count++;
+		print_esl_info(curr_esl.esl);
+
+		union {
+			const uint8_t *raw;
+			sv_esd_t *esd;
+		} curr_esd;
+		size_t esd_data_size;
+		uuid_t esd_owner;
+		enum signature_type sig_type = get_signature_type(curr_esl.esl->signature_type);
+
+		rc = next_esd_from_esl(curr_esl.raw, &curr_esd.raw, &esd_data_size, &esd_owner);
+		if (rc) {
+			prlog(PR_ERR, "Error reading esd %zu, rc = %d\n", count, rc);
+			return rc; 
 		}
 
-		print_esl_info(sig_list);
-		next_esl_size = esl_size;
-		offset = sizeof(sv_esl_t) + cpu_to_le32(sig_list->signature_header_size);
-		esl_size = esl_size - offset;
-		sig_offset = 0;
-
-		/* reads the esd from esl */
-		while (esl_size > 0) {
-			size_t data_size;
-
-			esd = (sv_esd_t *)(esl_data + (offset + sig_offset));
-			data_size = signature_size - sizeof(sv_esd_t);
-			cert = esd->signature_data;
-			cert_size = data_size;
-
+		while (curr_esd.esd != NULL) {
 			switch (sig_type) {
 			case ST_HASHES_START ... ST_X509_HASHES_END:
 				printf("\tData-%zu: ", count);
-				print_hex(cert, cert_size);
+				print_hex(curr_esd.esd->signature_data, esd_data_size);
```

Comment:
Not sure if this should be the `signature_data` printing here or just the whole ESD, but I have no files to test with to get expected output.

Reply:
Wanted to test how this goes but it turns out `test/guest/testenv/dbx/data` is an ESL with a cert in it:
```
[nchild@tinkpad secvarctl]$ ./bin/secvarctl -m guest read -p test/guest/testenv/ -n dbx 
READING dbx :
	ESL SIG LIST SIZE: 857
	GUID is : a159c0a5e494a74a87b5ab155c2bf072
	Signature type is: X509
	Certificate-1: 	Found certificate info
 Certificate:
    Data:
        Version: 3 (0x2)
        Serial Number:
            ea:4d:46:f1:fd:93:30:f3:54:ec:dd:02:19:00:54:3d:3f:1e:cb
        Signature Algorithm: sha256WithRSAEncryption
        Issuer:
            countryName               = NC
            organizationName          = testing corp
        Validity
            Not Before: Apr  7 09:18:56 2023 GMT
            Not After : May  7 09:18:56 2023 GMT
        Subject:
            countryName               = NC
            organizationName          = testing corp
        X509v3 extensions:
            X509v3 Subject Key Identifier: 
                61:3A:96:F3:36:CF:40:53:46:25:38:49:66:F3:CF:BF:38:18:C9:1F
            X509v3 Authority Key Identifier: 
                keyid:61:3A:96:F3:36:CF:40:53:46:25:38:49:66:F3:CF:BF:38:18:C9:1F

            X509v3 Basic Constraints: critical
                CA:TRUE
 
	Found 1 ESL's

RESULT: SUCCESS
```
I am not 100% sure whats supported so whatever. 

I then went to go generated a esl with a hash and hit more bugs so opening a PR shortly. Anyways  for brevity jumping over that dilemma.

When I finally got a generated ESL w/ hash, I could test and looks like we are overreadding:
```
[nchild@tinkpad secvarctl]$ hexdump guest.dbx
0000000 1626 c1c4 504c 4092 a9ac f941 9336 2843
0000010 004c 0000 0000 0000 0030 0000 0000 0000
0000020 0000 0000 0000 0000 0000 0000 ea98 4f6e
0000030 6f21 b42f 9fb6 9bff 443a 2c84 6838 a66c
0000040 f385 5df5 8cc4 3f5d 10b1 e47b          
000004c

[nchild@tinkpad secvarctl]$ ./bin/secvarctl -m guest read -e guest.dbx
	ESL SIG LIST SIZE: 76
	GUID is : 2616c4c14c509240aca941f936934328
	Signature type is: SHA256
	Data-1: 38:68:6c:a6:85:f3:f5:5d:c4:8c:5d:3f:b1:10:7b:e4:00:00:00:00:00:00:00:00:00:00:00:00:11:04:00:00
	Found 1 ESL's

RESULT: SUCCESS

[nchild@tinkpad secvarctl]$ ./bin/secvarctl-dbg -m guest read -e guest.dbx
	ESL SIG LIST SIZE: 76
	GUID is : 2616c4c14c509240aca941f936934328
	Signature type is: SHA256
=================================================================
==60409==ERROR: AddressSanitizer: heap-buffer-overflow on address 0x60700000006c at pc 0x000000403413 bp 0x7ffe916eea50 sp 0x7ffe916eea40
READ of size 1 at 0x60700000006c thread T0
    #0 0x403412 in print_hex /home/nchild/IBM/secvarctl-stuff/erichte-guest-devel/secvarctl/generic.c:39
    #1 0x44fc00 in print_variables backends/guest/common/read.c:165
    #2 0x452a1b in read_esl backends/guest/guest_svc_read.c:156
    #3 0x456067 in guest_read_command backends/guest/guest_svc_read.c:452
    #4 0x4069b4 in main /home/nchild/IBM/secvarctl-stuff/erichte-guest-devel/secvarctl/secvarctl.c:249
    #5 0x7fb7c44d3d84 in __libc_start_main (/lib64/libc.so.6+0x3ad84)
    #6 0x4031dd in _start (/home/nchild/IBM/secvarctl-stuff/erichte-guest-devel/secvarctl/bin/secvarctl-dbg+0x4031dd)

0x60700000006c is located 0 bytes to the right of 76-byte region [0x607000000020,0x60700000006c)
allocated by thread T0 here:
    #0 0x7fb7c5b59bb8 in __interceptor_malloc (/lib64/libasan.so.5+0xefbb8)
    #1 0x403b6e in get_data_from_file /home/nchild/IBM/secvarctl-stuff/erichte-guest-devel/secvarctl/generic.c:89
    #2 0x455d44 in guest_read_command backends/guest/guest_svc_read.c:440
    #3 0x4069b4 in main /home/nchild/IBM/secvarctl-stuff/erichte-guest-devel/secvarctl/secvarctl.c:249
    #4 0x7fb7c44d3d84 in __libc_start_main (/lib64/libc.so.6+0x3ad84)

SUMMARY: AddressSanitizer: heap-buffer-overflow /home/nchild/IBM/secvarctl-stuff/erichte-guest-devel/secvarctl/generic.c:39 in print_hex
Shadow bytes around the buggy address:
  0x0c0e7fff7fb0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0c0e7fff7fc0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0c0e7fff7fd0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0c0e7fff7fe0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0c0e7fff7ff0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
=>0x0c0e7fff8000: fa fa fa fa 00 00 00 00 00 00 00 00 00[04]fa fa
  0x0c0e7fff8010: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c0e7fff8020: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c0e7fff8030: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c0e7fff8040: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c0e7fff8050: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
==60409==ABORTING
```
Host backend can parse this properly:
```
[nchild@tinkpad secvarctl]$ ./bin/secvarctl -m host validate -e guest.dbx -x -v
----opening guest.dbx is success: reading 76 bytes----
VALIDATING ESL:
	ESL SIG LIST SIZE: 76
	GUID is : 2616c4c14c509240aca941f936934328
	Signature type is: SHA256
	Hash: 98:ea:6e:4f:21:6f:2f:b4:b6:9f:ff:9b:3a:44:84:2c:38:68:6c:a6:85:f3:f5:5d:c4:8c:5d:3f:b1:10:7b:e4
	Found 1 ESL's

RESULT: SUCCESS
```
I was able to get the outputs to line up and stop ASAN complaints by making the following change:
```
diff --git a/backends/guest/common/read.c b/backends/guest/common/read.c
index be32ad8..1712e81 100644
--- a/backends/guest/common/read.c
+++ b/backends/guest/common/read.c
@@ -162,7 +162,7 @@ int print_variables(const uint8_t *buffer, size_t buffer_size, const char *var_n
                        switch (sig_type) {
                        case ST_HASHES_START ... ST_X509_HASHES_END:
                                printf("\tData-%zu: ", count);
-                               print_hex(curr_esd.esd->signature_data, esd_data_size);
+                               print_hex((unsigned char *)curr_esd.raw, esd_data_size);
                        case ST_X509:
                                x509 = crypto_x509_parse_der(curr_esd.raw, esd_data_size);
```
I needed to typecast because we are dropping the `const` to call `print_hex`, obviously you can opt to just modify the definition of `print_hex` to include const, also with this change all uses of `curr_edt` union only use one implementation `curr_edt.raw`

Reply:
The arg to `print_hex` should probably be `const` anyway, so I will make that change. I'm not terribly married to the unions (pun...intended?) but I would like to overall reduce unnecessary casts in this code base.